### PR TITLE
[IOTDB-3815] replace remove data node to procedure

### PR DESCRIPTION
**Pull request by wangchao316**
replace remove data node to procedure

IoTDB> show cluster
show cluster
+------+----------+-------+---------+-----+
|NodeID|  NodeType| Status|     Host| Port|
+------+----------+-------+---------+-----+
|     0|ConfigNode|Running|  0.0.0.0|22277|
|     1|  DataNode|Running|127.0.0.1| 9003|
|     2|  DataNode|Running|127.0.0.1| 9005|
|     3|  DataNode|Running|127.0.0.1| 9007|
+------+----------+-------+---------+-----+
Total line number = 4

 remove-datanode.sh 127.0.0.1:6667
IoTDB> show cluster
show cluster
+------+----------+-------+---------+-----+
|NodeID|  NodeType| Status|     Host| Port|
+------+----------+-------+---------+-----+
|     0|ConfigNode|Running|  0.0.0.0|22277|
|     2|  DataNode|Running|127.0.0.1| 9005|
|     3|  DataNode|Running|127.0.0.1| 9007|
+------+----------+-------+---------+-----+
Total line number = 3
It costs 0.011s
IoTDB>

### CODE REVIEW

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java [hunk 1]
```diff
@@ -197,30 +197,38 @@ public TSStatus registerDataNode(RegisterDataNodePlan registerDataNodePlan) {
...
   /**
-   * Persist Infomation about remove dataNode
+   * Persist Information about remove dataNode
    *
    * @param req RemoveDataNodeReq
    * @return TSStatus
    */
   public TSStatus removeDataNode(RemoveDataNodePlan req) {
-    TSStatus result = new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
-    removeNodeInfo.removeDataNode(req);
```
Comment (BanDuWuZi) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java:207]: the inner class RemoveNodeInfo is useless, the state of RemoveDataNodePlan is persient by Porcudere

Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java:207]: fixed

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+            .filter(loc -> !loc.equals(disabledDataNode))
+            .map(TDataNodeLocation::getInternalEndPoint)
+            .collect(Collectors.toList());
+
+    for (TEndPoint server : otherOnlineDataNodes) {
+      TDisableDataNodeReq disableReq = new TDisableDataNodeReq(disabledDataNode);
+      status =
+          SyncDataNodeClientPool.getInstance()
+              .sendSyncRequestToDataNodeWithRetry(
+                  server, disableReq, DataNodeRequestType.DISABLE_DATA_NODE);
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java:99]: Would it be better if we use AsyncDataNodeClientPool?

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+  public TSStatus broadcastDisableDataNode(TDataNodeLocation disabledDataNode) {
+    LOGGER.info(
+        "DataNodeRemoveService start send disable the Data Node to cluster, {}", disabledDataNode);
+    TSStatus status = new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
+    List<TEndPoint> otherOnlineDataNodes =
+        configManager.getLoadManager().getOnlineDataNodes(-1).stream()
+            .map(TDataNodeConfiguration::getLocation)
+            .filter(loc -> !loc.equals(disabledDataNode))
+            .map(TDataNodeLocation::getInternalEndPoint)
+            .collect(Collectors.toList());
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java:92]: It seems that the annotation is not match with the inner code logic. By the way, why should the other onlineDataNodes need to know that there is a DataNode is removed?

**Review by CRZbulabula — COMMENTED** [review]
PTAL~

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java (hunk 2 above)
Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java:99]: If an asynchronous operation fails, the operation after the interruption cannot be ignored.

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java (hunk 3 above)
Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java:92]: Other datanodes do not send requests to the datanodes to be deleted in a timely manner.

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+  /**
+   * Get all consensus group id in this node
+   *
+   * @param dataNodeLocation data node location
+   * @return group id list
+   */
+  public List<TConsensusGroupId> getDataNodeRegionIds(TDataNodeLocation dataNodeLocation) {
+    return configManager.getPartitionManager().getAllReplicaSets().stream()
+        .filter(rg -> rg.getDataNodeLocations().contains(dataNodeLocation))
+        .filter(rg -> rg.regionId.getType() != TConsensusGroupType.PartitionRegion)
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java:72]: Will put two filter conditions into one be more efficiently?

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java (hunk 3 above)
Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java:92]: This logic is the same as the previous logic. I just added it to Procedure, and then changed it later if necessary.

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java (hunk 2 above)
Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java:99]: This logic is the same as the previous logic. I just added it to Procedure, and then changed it later if necessary.

**Comment by Beyyes** [comment]
image
The port 6667 doesn't appears in the datanodes

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/scheduler/LockQueue.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.ArrayDeque;
+
+/** Lock Queue for procedure of the same type */
+public class LockQueue {
+  private final ArrayDeque<Procedure> deque = new ArrayDeque<>();
+
+  private Procedure<?> lockOwnerProcedure = null;
+
+  public boolean tryLock(Procedure<?> procedure) {
```
[the remark this answers is not in the document]
Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/scheduler/LockQueue.java:32]: I use configNodeProcedureEnv.getSchedulerLock().lock()  in AbstractNodeProcedure. so this is no problem.

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java [hunk 6]
```diff
@@ -197,30 +189,38 @@ public TSStatus registerDataNode(RegisterDataNodePlan registerDataNodePlan) {
...
-    return result;
+    try {
+      dataNodeInfoReadWriteLock.writeLock().lock();
+      req.getDataNodeLocations()
+          .forEach(
+              drainingDataNodes -> {
```
[the remark this answers is not in the document]
Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/NodeInfo.java:202]: we support remove mult node once,  so req contains all of removed node.  I will rename drainingDataNodes  to removeDataNodes

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/RemoveDataNodeProcedure.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+/** remove config node procedure */
```
[the remark this answers is not in the document]
Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/RemoveDataNodeProcedure.java]: fixed

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java (hunk 4 above)
Comment (wangchao316) [reply confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java]: fixed

**Comment by wangchao316** [comment]
127.0.0.1:6667 , this is in windows, pseudo-distributed

File: server/src/main/java/org/apache/iotdb/db/service/DataNodeServerCommandLine.java [hunk 8]
```diff
@@ -75,10 +92,111 @@ protected int run(String[] args) {
...
+      removeNodesFromCluster(args);
+      removeTail();
+    } catch (Exception e) {
+      logger.error("remove Data Nodes error", e);
+    }
+  }
+
+  private void removePrepare(String[] args) throws BadNodeUrlException {
+    ConfigNodeInfo.getInstance()
+        .updateConfigNodeList(IoTDBDescriptor.getInstance().getConfig().getTargetConfigNodeList());
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/service/DataNodeServerCommandLine.java:119]: Is it necessary to updateConfigNodeList here?

**Review by CRZbulabula — COMMENTED** [review]
PTAL~

**Review by CRZbulabula — APPROVED** [review]
LGTM～

File: server/src/main/java/org/apache/iotdb/db/service/DataNodeServerCommandLine.java (hunk 8 above)
Comment (wangchao316) [reply server/src/main/java/org/apache/iotdb/db/service/DataNodeServerCommandLine.java:119]: necessary

**Review by Beyyes — APPROVED** [review]

**Pull request merged**